### PR TITLE
Implement Least Core

### LINKED ISSUE #157: Implement least core
Implement Least Core as described in [If You Like Shapley Then You’ll Love the Core ](https://ojs.aaai.org/index.php/AAAI/article/view/16721) to reproduce all of: 
Wang, Tianhao, Yu Yang, and Ruoxi Jia. ‘Improving Cooperative Game Theory-Based Data Valuation via Data Utility Learning’. arXiv, 2022. https://doi.org/10.48550/arXiv.2107.06336.

### PR DESCRIPTION
**Pull request by AnesBenmerzoug**
### Description

This PR closes #157 

### Changes

- Creates least_core subpackage
- Creates naive module inside that subpackage with the implementation of the exact least core method.
- Creates montecarlo module inside that subpackage with the implementation of the montecarlo least core method.
- Creates unit tests for both methods using the Mining Problem described in the [Core's Wikipedia page](https://en.wikipedia.org/wiki/Core_(game_theory)).
- Adds notebook demonstrating usage of Least Core on a synthetic dataset.
- Adds a `from_arrays()` class method to the `Dataset` and `GroupedDataset` classes. It makes it simple to create a dataset from sklearn's synthetic dataset function `make_*`
- Adds `compute_removal_score()` and `compute_random_removal_score()`, two helper functions used in the least core notebook to compute the accuracy after removing a percentage of data points, to the `pydvl.reporting.scores` module.
- Adds `extra_values` argument to `ValuationResult`

### Checklist

- [X] Wrote Unit tests (if necessary)
- [x] Updated Documentation (if necessary)
- [x] Updated Changelog
- [x] If notebooks were added/changed, added boilerplate cells are tagged with `"nbsphinx":"hidden"`

### CODE REVIEW

File: notebooks/least_core_basic.ipynb [hunk 1]
```diff
@@ -0,0 +1,811 @@
...
+    "The Least Core is the solution to the following linear programming problem:\n",
+    "\n",
+    "$$\n",
+    "\\begin{array}{lll}\n",
+    "\\text{minimize} & \\displaystyle{e} & \\\\\n",
+    "\\text{subject to} & \\displaystyle\\sum_{i\\in N} x_{i} = v(N) & \\\\\n",
+    "& \\displaystyle\\sum_{i\\in S} x_{i} + e \\geq v(S) &, \\forall S \\subseteq N \\\\\n",
+    "\\end{array}\n",
+    "$$\n",
+    "\n",
```
Comment (Xuzzo) [inline notebooks/least_core_basic.ipynb:21]: Can you add a few more words on how these equations are related to data valuation? Just an intuitive idea, if possible.

File: notebooks/least_core_basic.ipynb [hunk 2]
```diff
@@ -0,0 +1,811 @@
...
+    }
+   ],
+   "source": [
+    "%%time\n",
+    "budget_array = np.linspace(200, 2 ** (len(small_dataset) - 1), num=10, dtype=int)\n",
+    "\n",
+    "all_estimated_values_df = []\n",
+    "all_errors = {}\n",
+    "\n",
+    "for budget in budget_array:\n",
```
Comment (Xuzzo) [inline notebooks/least_core_basic.ipynb:400]: can you put a loading bar in the cells that take long? Mostly, those that take >1 minute. Basically, all those that loop over budget array

File: notebooks/least_core_basic.ipynb [hunk 3]
```diff
@@ -0,0 +1,811 @@
...
+   "cell_type": "markdown",
+   "id": "1e7d81aa",
+   "metadata": {},
+   "source": [
+    "## Dataset\n",
+    "\n",
+    "We will use a synthetic dataset. We sample 200 data points from a 50-dimensional Gaussian distribution.\n",
+    "\n",
+    "The dataset is generated using the [`make_classification`](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.make_classification.html) function from scikit-learn."
+   ]
```
Comment (Xuzzo) [inline notebooks/least_core_basic.ipynb:140]: I think it is helpful to add a small description of the dataset here. Just a few words to understand how to increase/decrease complexity if user wants to experiment

File: notebooks/least_core_basic.ipynb [hunk 4]
```diff
@@ -0,0 +1,811 @@
...
...Q1pQkm/pqNrSfbWVE6sUt4zrEbqVDSWH6Z7yr6wq86pMjAsqiKkUlzrGmULMcyYubxcOLdERQbpEmMwT9ueIkjnC/gKK9edTUjNGGk7RqT7O4oh4lASTZOH4gxUrN6CNoI2FQwh4HpaTbDrTPipkbXOmfsPAJe8GpFm8bprbQ07C7vKvLJpCYFU2J+3WSDs1knadpFUla1dYJ6ApITpVdFwbe2pppHytEJLCZvLIDXgUDqOTogTORbST8yR27OSXb50AHCH9fJJ+PolRaelsWiidTYp8K7HJuQ1Ooi1K0foj4lF+D8FoVBBqVEbcRSMOo8qIw+hhVlePCes6znI4HWftBpfVelftVuLWhmnlviDvhJu60+62YYcy25yk2Dhti3JyH85+EEjhQFKuPNE2KGcrnfAuwEqElRBxBodBxDzQkw5ZGmIee/i3RUd1FHoRuDFy+ybwqQ3z/JfAv1RK/adAHfiJkcd+dcNjL258AqXUXwf+OsCVK1cOZKU9jzYiwmt3rvLO4k2mm6cf9OrsCis5N+y1ssvbG7yXXyUlBeCCvsRnos/zVPAsTwbPUFM76GYwEJZaBpUrJBBcdFQ7x7IMbn6BZGUZhcJUq4R7tqRLma80X+YraXJbuW++kssVC98uOsIli1XiyR6P/ewPmProzJYd4UapJLe5PPdPaPau0o0v8u65v0Kn+sTmNRNIxOLEYZQmUBrjD4rvi6gAawKsuYc4uueFD/KvsvWC1QbX1UZRSw3msUlxn00x/TaaHMN6Z9ZGsnCcNJqkbS7Qbz5HFp0mjU+XDqQJRB8fAelBMlpOEhyCdDk8G3+f7JP15SUbBjN9hesrslWNy4oBjc0NzgaIbPwcHVHYJgpXGQtniE6vEkerROEqcbhCFLVRgSrcFcZAomDewLIup224HF43W0zbeGn27AB9EPgSuGOCc6g0RdKELMnI0wpBagmyFkE2h0oSSCzSd5BYyLYRmhXry84atc2laJUNDiO/b3zkUJqiG+VhnbBwat12fP1Jintv70dPXBSO2xHha4/5V56HE/Oj2+eyPiwcp6PUXwX+noj8f5RSPwz8llLqgzt9sIj8HeDvALzwwgteFvbsm7fnb/DqzFWmx6ZQx/xAxorllr3O27ZwKL2bv01KUaZ1Xl/kk9HneDp4lifNM9T1Lhw/AiZRxK0AlZXCUnw0Py9xlmy1
```
Comment (Xuzzo) [inline notebooks/least_core_basic.ipynb:785]: In the current plot it looks like the highest budget actually leads to lower accuracy, right? I wonder if this is related to the fact that the dataset is quite simple. It feels like more complex data would have scores more dependent on budget.

**Review by Xuzzo — COMMENTED** [review]
Code looks good! I feel like notebook is missing a bit of context: what is the main appeal of least core compared to Shapley? Is it faster? Even just a small comment would help. Maybe at some point we can also do a thorough comparison of performance given same budget.

File: notebooks/least_core_basic.ipynb (hunk 2 above)
Comment (AnesBenmerzoug) [reply notebooks/least_core_basic.ipynb:400]: I didn't want to add progress bars there because their output is not saved and they would show up in the documentation as having 0 progress.

I would rather add a note above those cells to warn the readers about those cells taking a long time to finish.

File: notebooks/least_core_basic.ipynb (hunk 3 above)
Comment (AnesBenmerzoug) [reply notebooks/least_core_basic.ipynb]: I added some more information to that description.

File: notebooks/least_core_basic.ipynb (hunk 4 above)
Comment (AnesBenmerzoug) [reply notebooks/least_core_basic.ipynb]: Yeah, you're right. I changed the description after that plot to describe two reasons why this could be the case.

File: notebooks/least_core_basic.ipynb (hunk 1 above)
Comment (AnesBenmerzoug) [reply notebooks/least_core_basic.ipynb]: Done. I also added a description of the difference between Shapley and the Core to both the notebook and the docs.

File: notebooks/least_core_basic.ipynb (hunk 2 above)
Comment (AnesBenmerzoug) [reply notebooks/least_core_basic.ipynb]: On second thought, you're right the progress bars make more sense than a note.

File: notebooks/least_core_basic.ipynb (hunk 1 above)
Comment (Xuzzo) [reply notebooks/least_core_basic.ipynb]: there is a typo at the beginning of the notebook =)

Comment (AnesBenmerzoug) [reply notebooks/least_core_basic.ipynb]: I fixed it. Thanks!

**Review by mdbenito — APPROVED** [review]
Great (and lots of) work! I've reviewed everything but the notebook, which I just scanned superficially because it will probably grow and change soon.

File: docs/30-data-valuation.rst [hunk 5]
```diff
@@ -324,13 +324,127 @@ and can be used in pyDVL with:
...
+
+$$
+P_{S\sim D}\left[\sum_{x_i\in S} v_u(x_i) + e^{*} + \epsilon \geq v_u(S)\right] \geq 1 - \delta
+$$
+
+Where $e^{*}$ is the optimal least core value.
+
+With these relaxations, we obtain a polynomial running time.
+
+Using Owen sampling follows the same pattern as every other method for Shapley
```
[the remark this answers is not in the document]
Comment (AnesBenmerzoug) [reply docs/30-data-valuation.rst:420]: Oups...

File: src/pydvl/utils/dataset.py [hunk 6]
```diff
@@ -155,15 +159,11 @@ def get_training_data(self, train_indices: Optional[Iterable[int]]):
...
-    def get_test_data(self, test_indices: Optional[Iterable[int]]):
-        """Given a set of indices, returns the test data that refer to those
-        indices."""
-        if test_indices is None:
-            return self.x_test, self.y_test
-        else:
-            x = self.x_test[test_indices]
-            y = self.y_test[test_indices]
-            return x, y
+    def get_test_data(
```
[the remark this answers is not in the document]
Comment (AnesBenmerzoug) [reply src/pydvl/utils/dataset.py:162]: I kept the argument because I wanted to keep the interface the same as for `get_training_data()`.

For data valuation we always return the entire test set, so we can safely ignore the indices. But for feature valuation we need in the indices and for that I can just overwrite the `get_test_data()` method in a subclass and the Utility remains the same.

File: src/pydvl/utils/parallel/backend.py [hunk 7]
```diff
@@ -137,7 +138,7 @@ def __init__(self, config: ParallelConfig):
         self.config = config_dict
         if self.config["address"] is None:
             self.config["ignore_reinit_error"] = True
-        ray.init(**self.config)
+        ray.init(logging_level=logging.WARNING, **self.config)
```
[the remark this answers is not in the document]
Comment (AnesBenmerzoug) [reply src/pydvl/utils/parallel/backend.py:141]: True. I will add that to ParallelConfig

File: src/pydvl/utils/utility.py [hunk 8]
```diff
@@ -139,10 +137,11 @@ def _utility(self, indices: FrozenSet) -> float:
         if not indices:
             return 0.0
 
-        x, y = self.data.get_training_data(list(indices))
+        x_train, y_train = self.data.get_training_data(list(indices))
+        x_test, y_test = self.data.get_test_data(list(indices))
```
[the remark this answers is not in the document]
Comment (AnesBenmerzoug) [reply src/pydvl/utils/utility.py:141]: Yes, but as I said above I wanted to keep the interface the same for the feature valuation experiments.

File: src/pydvl/value/least_core/_common.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+        b_eq=b_eq,
+        bounds=bounds,
+        method="highs-ipm",
+        options=options,
+    )
+
+    logger.debug(f"{result=}")
+
+    if result.success:
+        values = np.asarray(result.x)
```
[the remark this answers is not in the document]
Comment (AnesBenmerzoug) [reply src/pydvl/value/least_core/_common.py:60]: True. I will remove the else and the extra level of indentation

File: docs/30-data-valuation.rst [hunk 10]
```diff
@@ -324,13 +324,127 @@ and can be used in pyDVL with:
...
+
+- **Coalitional rationality**:
+  The sum of payoffs to the agents in any coalition S is at
+  least as large as the amount that these agents could earn by
+  forming a coalition on their own.
+  $$\displaystyle\sum_{x_i\in S} v_u(x_i) \geq v_u(S), \forall S \subseteq D\,$$
+
+The second property states that the sum of payoffs to the agents in any subcoalition S is at
+least as large as the amount that these agents could earn by
+forming a coalition on their own.
```
[the remark this answers is not in the document]
Comment (AnesBenmerzoug) [reply docs/30-data-valuation.rst:356]: I would prefer leaving them in for now and updating the docs with links to the review once it's published.

File: src/pydvl/reporting/scores.py [hunk 11]
```diff
@@ -156,3 +159,67 @@ def compute_fb_scores(
     }
 
     return results
+
+
+def compute_removal_score(
```
[the remark this answers is not in the document]
Comment (AnesBenmerzoug) [reply src/pydvl/reporting/scores.py:164]: True. Those are old functions that are not used anywhere. I will remove them.

File: src/pydvl/utils/utility.py [hunk 12]
```diff
@@ -13,19 +13,17 @@
...
 import numpy as np
+from numpy.typing import NDArray
 from sklearn.metrics import check_scoring
 
 from .caching import CacheStats, memcached, serialize
 from .config import MemcachedConfig
 from .dataset import Dataset
 from .types import Scorer, SupervisedModel
 
-if TYPE_CHECKING:
```
[the remark this answers is not in the document]
Comment (AnesBenmerzoug) [reply src/pydvl/utils/utility.py:26]: I started it and then it was just carried on... sorry

File: src/pydvl/utils/utility.py [hunk 13]
```diff
@@ -226,9 +225,9 @@ def __init__(
         self.model = model
         self._current_iteration = 0
         self._is_model_fit = False
-        self._utility_samples: Dict[FrozenSet, Tuple["NDArray", float]] = {}
+        self._utility_samples: Dict[FrozenSet, Tuple[NDArray, float]] = {}
 
-    def _convert_indices_to_boolean_vector(self, x: Iterable[int]) -> "NDArray":
+    def _convert_indices_to_boolean_vector(self, x: Iterable[int]) -> NDArray:
```
[the remark this answers is not in the document]
Comment (AnesBenmerzoug) [reply src/pydvl/utils/utility.py]: It should be `np.bool_`

File: docs/30-data-valuation.rst [hunk 14]
```diff
@@ -324,13 +324,127 @@ and can be used in pyDVL with:
    utility = Utility(model, data)
    values = compute_shapley_values(u=utility, mode="knn")
 
+.. _Least Core:
```
[the remark this answers is not in the document]
Comment (AnesBenmerzoug) [reply docs/30-data-valuation.rst:336]: Done. Could you please tell me the reason for this?

[the remark this answers is not in the document]
Comment (mdbenito) [reply docs/30-data-valuation.rst:336]: * Diffs easier to handle.
* Research shows that eye fatigue is reduced with shorter line lengths.
* 2 vertical panes in FullHD small screens, 3 in UHD larger ones
* Convention has to be set at some number.

File: docs/30-data-valuation.rst [hunk 15]
```diff
@@ -324,13 +324,127 @@ and can be used in pyDVL with:
...
+- **Efficiency**:
+  The payoffs are distributed such that it is not possible to make any participant better off
+  without making another one worse off.
+  $$\displaystyle\sum_{x_i\in D} v_u(x_i) = v_u(D)\,$$
+
+- **Coalitional rationality**:
+  The sum of payoffs to the agents in any coalition S is at
+  least as large as the amount that these agents could earn by
+  forming a coalition on their own.
+  $$\displaystyle\sum_{x_i\in S} v_u(x_i) \geq v_u(S), \forall S \subseteq D\,$$
```
[the remark this answers is not in the document]
Comment (mdbenito) [reply docs/30-data-valuation.rst]: This is different from the review, but I agree that internal consistency is more important. At any rate, if $v_u$ denotes value, then the RHS should be $U(S)$, shouldn't it?

File: src/pydvl/reporting/scores.py [hunk 16]
```diff
@@ -156,3 +159,67 @@ def compute_fb_scores(
...
+
+    for pct in percentages:
+        n_removal = int(pct * len(u.data))
+        indices = values.indices[n_removal:]
+        score = u(indices)
+        scores[pct] = score
+    return scores
+
+
+def compute_random_removal_score(
```
[the remark this answers is not in the document]
Comment (mdbenito) [reply src/pydvl/reporting/scores.py:82]: yes. But it's just an idea.

File: src/pydvl/utils/dataset.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (mdbenito) [reply src/pydvl/utils/dataset.py:162]: I'm not sure I get the need for feature valuation, so I'll let you be the judge of this. However, you have the same behaviour if you leave the possibility of returning just a few indices or all of them if the arg is `None`.

File: src/pydvl/utils/utility.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (mdbenito) [reply src/pydvl/utils/utility.py:141]: I see, it's a bit weird, so maybe it deserves documentation (not only a statement of behaviour but an explanation of why this is so)

File: src/pydvl/utils/dataset.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (mdbenito) [reply src/pydvl/utils/dataset.py:162]: See my comment below.

**Review by mdbenito — APPROVED** [review]
I don't have any more suggestions, other than documenting the reason for get_test_data's arguments

File: docs/30-data-valuation.rst (hunk 14 above)
[the remark this answers is not in the document]
Comment (AnesBenmerzoug) [reply docs/30-data-valuation.rst:336]: Thanks for the explanation.

File: docs/30-data-valuation.rst (hunk 15 above)
[the remark this answers is not in the document]
Comment (AnesBenmerzoug) [reply docs/30-data-valuation.rst]: You're right. I will change that. We can update the documentation once the review is out.

File: src/pydvl/utils/utility.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (AnesBenmerzoug) [reply src/pydvl/utils/utility.py:141]: You're right. I will add an explanation for that behaviour  in the case of data valuation and give an example of what could be done to modify this.

File: src/pydvl/reporting/scores.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (AnesBenmerzoug) [reply src/pydvl/reporting/scores.py]: That's not a bad idea. I removed the nearly duplicate function and added a `from_random()` class method to `ValuationResult` that creates an instance with random values of a given size.

File: src/pydvl/utils/utility.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (AnesBenmerzoug) [reply src/pydvl/utils/utility.py:141]: Done.

**Comment by AnesBenmerzoug** [comment]
@mdbenito, @Xuzzo I will merge this hopefully tomorrow if there are no more changes to be made.

**Review by mdbenito — APPROVED** [review]

File: src/pydvl/reporting/scores.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (mdbenito) [reply src/pydvl/reporting/scores.py]: Go ahead and merge after that then
